### PR TITLE
CASSANDRA-20154: Fix BETWEEN bounds to preserve inverted bounds

### PR DESCRIPTION
**Pull request by arvindKandpal-ksolves**
Fix CASSANDRA-20154: BETWEEN operator returns wrong results when start bound is greater than end bound.

This PR fixes a bug where queries like `SELECT * FROM table WHERE column BETWEEN 5 AND 3` 
incorrectly return rows instead of an empty result set. The issue was caused by Cassandra 
automatically sorting BETWEEN bounds, which violates SQL standards.

**Changes made:**
- Modified SimpleRestriction.java to preserve user-specified bound order for BETWEEN queries
- Updated ClusteringColumnRestrictions.java to properly handle BETWEEN on first clustering column
- Added tests to verify inverted BETWEEN bounds now return empty results

**Before:** `WHERE pk BETWEEN 5 AND 3` would return rows between 3 and 5
**After:** `WHERE pk BETWEEN 5 AND 3` correctly returns no rows (as per SQL standard)

By Arvind Kandpal for CASSANDRA-20154

### CODE REVIEW

**Comment by arvindKandpal-ksolves** [comment]
Hi team!

This PR has been open for 4 days and I'd appreciate some feedback on my fix for **CASSANDRA-20154**.

**Quick Summary:**
- Fixes BETWEEN operator to handle inverted bounds correctly per SQL standards  
- `WHERE pk BETWEEN 5 AND 3` now returns empty results (was incorrectly returning rows)
- Changes in `SimpleRestriction.java` and `ClusteringColumnRestrictions.java`
- Added comprehensive tests to verify the fix

**Technical Details:**
The core issue was Cassandra automatically sorting BETWEEN bounds, causing inverted bounds like `BETWEEN 5 AND 3` to return rows instead of empty results. This PR preserves user-specified bound order to match SQL standard behavior.

**Testing:**
- Added tests for inverted BETWEEN bounds  
- Verified existing functionality remains intact  
- Covers both partition key and clustering column scenarios  

Would really appreciate review from the community! Particularly looking for feedback on:
1. Any edge cases I might have missed
2. Performance implications 
3. Additional test scenarios to consider

@dineshjoshi @maedhroz @smiklosovic @jmckenzie-dev  @ifesdjeen - would any of you have time to take a look when convenient? 

Thanks for your time!

**Review by dcapwell — COMMENTED** [review]
took a quick look and it looks like this doesn't remove all the sorts? `org.apache.cassandra.cql3.Operator.BETWEEN#restrict` still sorts

You can also best detect this issue by removing `org.apache.cassandra.cql3.KnownIssue#BETWEEN_START_LARGER_THAN_END`

you can remove that and run `org.apache.cassandra.distributed.test.cql3.SingleNodeTableWalkTest#test` and `org.apache.cassandra.distributed.test.cql3.SingleNodeTokenConflictTest#test` and will see this problem pop up.

File: src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java [hunk 1]
```diff
@@ -210,6 +216,17 @@ public void addToRowFilter(RowFilter filter,
...
+        {
+            SimpleRestriction sr = (SimpleRestriction) restriction;
+            if (sr.operator() == Operator.BETWEEN &&
+                    index == sr.firstColumn().position())
+            {
+                return true;
+            }
+        }
+        // Default: require filtering or non‑leading‑column position
+        return restriction.needsFilteringOrIndexing() ||index != restriction.firstColumn().position();
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java]: can you fix the spacing?

File: src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java [hunk 2]
```diff
@@ -210,6 +216,17 @@ public void addToRowFilter(RowFilter filter,
 
     private boolean handleInFilter(SingleRestriction restriction, int index)
     {
-        return restriction.needsFilteringOrIndexing() || index != restriction.firstColumn().position();
+        // Allow BETWEEN on the first clustering column to be handled via row filter
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java]: can you expand on this?  also if this is a slice (which uses `restrict`) how does this work?

File: src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java [hunk 3]
```diff
@@ -343,9 +343,12 @@ public void addToRowFilter(RowFilter filter, IndexRegistry indexRegistry, QueryO
                 List<ByteBuffer> buffers = bindAndGet(options);
                 if (operator.kind() != Operator.Kind.BINARY)
                 {
-                    // For BETWEEN we support like in SQL reversed bounds
-                    if (operator.kind() == Operator.Kind.TERNARY)
+                    // New: only sort IN‑list values; preserve order for BETWEEN
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java]: can remove `New:`

**Comment by dcapwell** [comment]
checked and sadly the AST tests don't look to cover the slice case.  I wouldn't push that on you as it isn't as simple as adding the test logic, its fixing anything it finds (which is 95% of the effort historically)

**Comment by arvindKandpal-ksolves** [comment]
Hi @dcapwell,

I have rebased against trunk and addressed the feedback:
**Operator.java:** Removed sorting logic to comply with SQL standards.
**ClusteringColumnRestrictions.java:** Reverted to trunk state (the Operator fix handles slices correctly now).
**Tests:** Removed the KnownIssue flag, updated existing tests, and added BetweenInversionTest.
Verified `SingleNodeTableWalkTest` and `SingleNodeTokenConflictTest` pass locally.

Ready for re-review!

File: test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests that the BETWEEN operator respects SQL semantics:
+ * - Normal BETWEEN (low <= high) returns rows.
+ * - Inverted BETWEEN (low > high) returns no rows (empty result).
+ */
+public class BetweenInversionTest
```
Comment (rajucomp) [inline test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java]: nit: public modifier not needed on tests or test methods

Comment (arvindKandpal-ksolves) [reply test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java]: @rajucomp I tried removing the public modifier, but the test runner failed locally with an initialization error. It seems the current setup requires test classes to be public, so I've kept it as is to ensure CI passes.

Comment (rajucomp) [reply test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java]: Could you post the error that you are getting ?

Comment (arvindKandpal-ksolves) [reply test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java]: _build-test:
    [javac] Compiling 1 source file to /mnt/backup_hdd/cassandra-github-code/cassandra/build/test/classes
    [javac] Note: /mnt/backup_hdd/cassandra-github-code/cassandra/test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java uses unchecked or unsafe operations.
    [javac] Note: Recompile with -Xlint:unchecked for details.
     [echo] Non default JDK version used: 17

_check-test-names:
[check-test-names] 17:34:35.847 [main] INFO org.reflections.Reflections -- Reflections took 4480 ms to scan 4 urls, producing 1384 keys and 16809 values
[check-test-names] 17:34:40.586 [main] INFO org.reflections.Reflections -- Reflections took 1227 ms to scan 4 urls, producing 1384 keys and 16809 values

test-jvm-dtest:
[junit-timeout] Testsuite: org.apache.cassandra.distributed.test.BetweenInversionTest-_jdk17
[junit-timeout] Testsuite: org.apache.cassandra.distributed.test.BetweenInversionTest-_jdk17 Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.131 sec
[junit-timeout] 
[junit-timeout] Testcase: initializationError(org.apache.cassandra.distributed.test.BetweenInversionTest)-_jdk17:	Caused an ERROR
[junit-timeout] The class org.apache.cassandra.distributed.test.BetweenInversionTest is not public.
[junit-timeout] java.lang.Exception: The class org.apache.cassandra.distributed.test.BetweenInversionTest is not public.
[junit-timeout] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[junit-timeout] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
[junit-timeout] 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[junit-timeout] 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
[junit-timeout] 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
[junit-timeout] 
[junit-timeout] 
[junit-timeout] Testcase: initializationError(org.apache.cassandra.distributed.test.BetweenInversionTest)-_jdk17:	Caused an ERROR
[junit-timeout] Test class should have exactly one public constructor
[junit-timeout] java.lang.Exception: Test class should have exactly one public constructor
[junit-timeout] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[junit-timeout] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
[junit-timeout] 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[junit-timeout] 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
[junit-timeout] 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
[junit-timeout] 
[junit-timeout] 
[junit-timeout] Test org.apache.cassandra.distributed.test.BetweenInversionTest FAILED

BUILD FAILED
/mnt/backup_hdd/cassandra-github-code/cassandra/build.xml:1850: The following error occurred while executing this line:
/mnt/backup_hdd/cassandra-github-code/cassandra/build.xml:1732: Some test(s) failed.

Total time: 1 minute 4 seconds

Comment (rajucomp) [reply test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java]: Thanks for providing the log. It seems weird to make the tests public. But I guess it is what it is.

**Comment by dcapwell** [comment]
patch mostly looks fine, trying to test it locally; specifically if i run the AST fuzz tests 1k+ times do they fail with this change (aka was something missed).

**Comment by dcapwell** [comment]
AST tests are failing

FullAccordInteropMultiNodeTableWalkTest: seed=-4109840071238703726

```
		162: UPDATE ks7.tbl SET s0 += [00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000] WHERE  pk0 = 'Py\u001C3\u0012|R' -- on node3
		163: SELECT * FROM ks7.tbl WHERE token(pk0) >= -9223372036854775808 AND token(pk0) < 9173422854434358258 PER PARTITION LIMIT 155 -- min token range, on node1, fetch size 1

	at accord.utils.Property$StatefulBuilder.check(Property.java:554)
	at org.apache.cassandra.distributed.test.cql3.SingleNodeTableWalkTest.test(SingleNodeTableWalkTest.java:369)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Caused by: java.lang.AssertionError: Unexpected results for query: SELECT * FROM ks7.tbl WHERE token(pk0) >= -9223372036854775808 AND token(pk0) < 9173422854434358258 PER PARTITION LIMIT 155
Caused by: java.lang.AssertionError: 
Possible column conflicts:
	Expected: ['Py3|R', -2.3040869E-29, [00000000-0000-1500-8400-000000000000, 00000000-0000-1700-be00-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000], {f0: [00000000-0000-4000-b200-000000000000], f1: [true], f2: ['ì«˜ì¨“ã…»é•”å¢šë¶®á½©æ›ä¼´æ·–']}]
	Diff (expected over actual):
s0 list<timeuuid>                                                                                                                                       
[00000000-0000-1500-8400-000000000000, 00000000-0000-1700-be00-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000]
[00000000-0000-1500-8400-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000, 00000000-0000-1700-be00-000000000000]

	Expected: ['Py3|R', Infinity, [00000000-0000-1500-8400-000000000000, 00000000-0000-1700-be00-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000], {f0: [00000000-0000-4b00-a600-000000000000], f1: [true], f2: ['ä³¡?æ”ä™®']}]
	Diff (expected over actual):
s0 list<timeuuid>                                                                                                                                       
[00000000-0000-1500-8400-000000000000, 00000000-0000-1700-be00-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000]
[00000000-0000-1500-8400-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000, 00000000-0000-1700-be00-000000000000]

Expected:
pk0                 | ck0            | s0                                                                                                                                                       | v0                                                                          
'Py\u001C3\u0012|R' | -2.3040869E-29 | [00000000-0000-1500-8400-000000000000, 00000000-0000-1700-be00-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000] | {f0: [00000000-0000-4000-b200-000000000000], f1: [true], f2: ['ì«˜ì¨“ã…»é•”å¢šë¶®á½©æ›ä¼´æ·–']}
'Py\u001C3\u0012|R' | Infinity       | [00000000-0000-1500-8400-000000000000, 00000000-0000-1700-be00-000000000000, 00000000-0000-1b00-aa00-000000000000, 00000000-0000-1700-ab00-000000000000] | {f0: [00000000-0000-4b00-a600-000000000000], f1: [true], f2: ['ä³¡?æ”ä™®']}      
'''.\u0012P./.'     | -1.9964244E-30 | [00000000-0000-1f00-bd00-000000000000, 00000000-0000-1200-9700-000000000000]                                                                             | {f0: [00000000-0000-4c00-8600-000000000000], f1: [true], f2: ['ïŒ€é°ƒ']}        
'.1a/'              | null           | [00000000-0000-1700-a500-000000000000, 00000000-0000-1b00-8400-000000000000, 00000000-0000-1300-8300-000000000000]                                       | null                                                                        
'R%'                | null           | [00000000-0000-1c00-a000-000000000000]                                                                                                                   | null                                                                        
'!\u0005L~
!'       | -2.4086384E38  | [00000000-0000-1900-8300-000000000000, 00000000-0000-1900-9e00-000000000000]                                                                             | {f0: [00000000-0000-4000-9800-000000000000], f1: [true], f2: ['ï³¦']}         
'!\u0005L~
!'       | 1.4466378E35   | [00000000-0000-1900-8300-000000000000, 00000000-0000-1900-9e00-000000000000]                                                                             | {f0: [00000000-0000-4300-a200-000000000000], f1: [true], f2: ['èµ£ì¶…å‰§è•–å™‡ïž„î¨›']}   
```

Only accord seems to be failing, so looks to be missing something

**Comment by dcapwell** [comment]
test results look good, [CASSANDRA-20828](https://issues.apache.org/jira/browse/CASSANDRA-20828); popped up but thats not related to this patch and a known bug with accord

**Comment by dcapwell** [comment]
Looks like `test/unit/org/apache/cassandra/cql3/validation/operations/SelectSingleColumnRelationTest.java` is failing.  So far all failures look to be that the tests expected `between 5 and 2` to return data, so we should either remove the test or update the test to expect 0 results

File: test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package org.apache.cassandra.distributed.test;
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/BetweenInversionTest.java:18]: missing the license so `ant rat-check` is failing.  All files must have the apache license added to them

**Comment by dcapwell** [comment]
Looks like several SAI tests failed

`org.apache.cassandra.index.sai.cql.DescClusteringRangeQueryTest`

Can you look into this?

**Comment by dcapwell** [comment]
`SelectMultiColumnRelationTest` is also failing, looks like the same invalid test assumption.  Can you fix?

**Comment by dcapwell** [comment]
From Caleb

> if it's a real bug, it would probably be in RowFilter , sai.plan.Expression , or DataResolver

**Comment by dcapwell** [comment]
took awhile but found the issue

```
diff --git a/src/java/org/apache/cassandra/cql3/Operator.java b/src/java/org/apache/cassandra/cql3/Operator.java
index 63b78a4091..89ff700310 100644
--- a/src/java/org/apache/cassandra/cql3/Operator.java
+++ b/src/java/org/apache/cassandra/cql3/Operator.java
@@ -788,8 +788,10 @@ public enum Operator
         public boolean isSatisfiedBy(AbstractType<?> type, ByteBuffer leftOperand, ByteBuffer rightOperand)
         {
             List<ByteBuffer> buffers = ListType.getInstance(type, false).unpack(rightOperand);
-            // We use compare instead of compareForCQL to deal properly with reversed clustering columns
-            return type.compare(leftOperand, buffers.get(0)) >= 0 && type.compare(leftOperand, buffers.get(1)) <= 0;
+            // Use unwrapped type for comparison to handle reversed clustering columns correctly.
+            // BETWEEN semantics are always in logical (non-reversed) order: value >= low AND value <= high.
+            AbstractType<?> unwrapped = type.unwrap();
+            return unwrapped.compare(leftOperand, buffers.get(0)) >= 0 && unwrapped.compare(leftOperand, buffers.get(1)) <= 0;
         }

         @Override
diff --git a/src/java/org/apache/cassandra/index/sai/plan/Expression.java b/src/java/org/apache/cassandra/index/sai/plan/Expression.java
index a5cf07e35d..6f8bac0bb3 100644
--- a/src/java/org/apache/cassandra/index/sai/plan/Expression.java
+++ b/src/java/org/apache/cassandra/index/sai/plan/Expression.java
@@ -246,8 +246,8 @@ public abstract class Expression
                 Value first = new Value(buffers.get(0), indexTermType);
                 Value second = new Value(buffers.get(1), indexTermType);

-                // SimpleRestriction#addToRowFilter() ensures correct bounds ordering, but SAI enforces a non-arbitrary
-                // ordering between IPv4 and IPv6 addresses, so correction may still be necessary.
+                // BETWEEN values are not sorted by SimpleRestriction to preserve semantics (BETWEEN 5 AND 2 is invalid).
+                // SAI enforces a non-arbitrary ordering between IPv4 and IPv6 addresses, so correction may still be necessary.
                 boolean outOfOrder = indexTermType.compare(first.encoded, second.encoded) > 0;
                 lower = new Bound(outOfOrder ? second : first, true);
                 upper = new Bound(outOfOrder ? first : second, true);
(END)
```

`Operator.isSatisfiedBy` was not correctly working when the column is reversed.  The other part of the diff is fixing a SAI comment as its no longer valid with this patch.

fuzz tests are passing without issue but i think this points out that slice queries are very under tested with BETWEEN, so tomorrow ill try to enhance the fuzz tests to make that better

**Comment by arvindKandpal-ksolves** [comment]
Hi @dcapwell, thanks for the feedback.

I have addressed the major points and verified them locally:

Operator Logic: Updated isSatisfiedBy to use type.unwrap() for correct handling of reversed columns.

Counters Bug: Fixed SimpleRestriction to skip sorting for Counter types (resolving the UnsupportedOperationException).

Tests & License: Added the missing license header to BetweenInversionTest and updated SelectSingleColumnRelationTest to expect empty results for inverted bounds.

Current Issue: I am facing some complex edge cases in `SelectMultiColumnRelationTest` regarding mixed-order clustering columns. The row counts are inconsistent with the new empty-result logic.

Could you please review these? Also, any advice on how to handle those specific mixed-order multi-column tests would be really helpful.

Thanks!

**Comment by dcapwell** [comment]
Sent a feedback branch; trying to make sure all tests are stable and passing

https://github.com/arvindKandpal-ksolves/cassandra/pull/1

Take a look and let me know if you have any concerns/feedback

**Comment by arvindKandpal-ksolves** [comment]
Hi @dcapwell,

I have merged your feedback branch and verified everything locally.
_Verification Results:_ I ran an extensive suite of tests to ensure stability, and all passed successfully:

org.apache.cassandra.cql3.** (Complete CQL test suite)
org.apache.cassandra.cql3.validation.operations.**
SelectMultiColumnRelationTest & SelectSingleColumnRelationTest
BetweenInversionTest & SingleNodeTableWalkTest (JVM dtests)

**Current State:** The branch currently has 15 commits (including your feedback commits).

**Question:** Should I squash these into a single commit now, or will you handle the squash-and-merge on your end ?

**Comment by dcapwell** [comment]
> Question: Should I squash these into a single commit now, or will you handle the squash-and-merge on your end ?

w/e the committer merges, they do the following

1) squash into a single commit
2) create a new commit message / author
3) update changes for anything user facing (like this)

So no worries =)

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 6]
```diff
@@ -1640,6 +1645,18 @@ private List<PrimaryKey> applyLimits(List<PrimaryKey> primaryKeys, Optional<Valu
         return primaryKeys;
     }
 
+    private boolean shouldReverse(Select.OrderBy orderBy)
+    {
+        for (var block : orderBy.ordered)
+        {
+            Symbol col = (Symbol) block.expression; //TOOD (coverage): do we support anything other than symbol?
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java]:
```suggestion
            Symbol col = (Symbol) block.expression; //TODO (coverage): do we support anything other than symbol?
```

sorry, this is my bug...

**Review by dcapwell — APPROVED** [review]
+1

**Comment by arvindKandpal-ksolves** [comment]
Thanks @dcapwell! I've applied the typo fix you suggested. Also, thanks for clarifying the squash process.

File: test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java [hunk 7]
```diff
@@ -95,6 +95,11 @@ public T get(int index)
         return values[index];
     }
 
+    public T get(T value)
+    {
+        return get(indexOf(value));
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java:100]: nit: I'm guessing the possibility `indexOf()` returns `-1` was taken into account here. `IOOBE` isn't the best exception, but throwing something more descriptive isn't the most important thing in the world for test code.

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 8]
```diff
@@ -1640,6 +1645,18 @@ private List<PrimaryKey> applyLimits(List<PrimaryKey> primaryKeys, Optional<Valu
         return primaryKeys;
     }
 
+    private boolean shouldReverse(Select.OrderBy orderBy)
+    {
+        for (var block : orderBy.ordered)
+        {
+            Symbol col = (Symbol) block.expression; //TODO (coverage): do we support anything other than symbol?
+            col = factory.clusteringColumns.get(col); // switch to table symbol so we know if its reversed or not
```
Comment (maedhroz) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:1653]: So we're just using the `expression` `Symbol` to get the has to look up the one from the table?

**Review by maedhroz — APPROVED** [review]

File: test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java (hunk 7 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java:100]:
```suggestion
        int idx = indexOf(value);
        if (idx == -1) throw new IllegalArgumentException("Unable to find value " + value);
        return get(idx);
```

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java (hunk 8 above)
Comment (dcapwell) [reply test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:1653]: right now Order By uses `Symbol` but I think i made it more generic in case we moved closer to how SQL works... but it should be a symbol 100% of the time today.

File: test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java (hunk 7 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java:100]: spoke in slack; ill ninja this as part of my merge

**Comment by dcapwell** [comment]
```
commit bad0a103437e745ab179cc26c3f27c5acd5ef3ba
Author: Arvind Kandpal <arvind.kandpal@ksolves.com>
Date:   Tue Jan 20 12:31:36 2026 -0800

    BETWEEN where token(Y) > token(Z) returns wrong answer

    patch by Arvind Kandpal; reviewed by Caleb Rackliffe, David Capwell for CASSANDRA-20154
```
